File: frontend/src/app/mod-calendar/models/calendar-event-full.ts
import { IINode, INode } from "../../mod-files-data/mod-files-data.module";
import { IUser, User } from "../../mod-userdata/mod-userdata.module";
import { CalendarEventCore, ICalendarEventCore } from "./calendar-event-core";

/**
 * 
 */
export interface ICalendarEventFull {
    core: ICalendarEventCore,
    reqPersons: IUser[],
    optPersons: IUser[],
    hashTags: string[],
    attachments: IINode[]
}

/**
 * 
 */
export class CalendarEventFull {

    /**
     * 
     * @param core 
     * @param persons 
     * @param hashTags 
     * @param attachments 
     */
    constructor(
        public core: CalendarEventCore,
        public reqPersons: User[],
        public optPersons: User[],
        public hashTags: string[],
        public attachments: INode[]) {
    }

    /**
     * 
     * @returns 
     */
    public static empty(): CalendarEventFull {
        return new CalendarEventFull(
            CalendarEventCore.empty(),
            new Array<User>(),
            new Array<User>(),
            new Array<string>(),
            new Array<INode>()
        );
    }

    /**
     * 
     * @returns 
     */
    public isEmpty(): boolean {
        return this.core.isEmpty();
    }

    /**
     * 
     * @param json 
     * @returns 
     */
    public static fromJSON(json: ICalendarEventFull): CalendarEventFull {

        return new CalendarEventFull(
            CalendarEventCore.fromJSON(json.core),
            json.reqPersons.map(person => User.fromJSON(person)),
            json.optPersons.map(person => User.fromJSON(person)),
            json.hashTags,
            json.attachments.map(attachment => INode.fromJSON(attachment))
        );
    }

    /**
     * 
     * @returns 
     */
    public toJSON(): ICalendarEventFull {

        return {
            core: this.core.toJSON(),
            reqPersons: this.reqPersons.map(person => person.toJSON()),
            optPersons: this.optPersons.map(person => person.toJSON()),
            hashTags: this.hashTags,
            attachments: this.attachments.map(attachment => attachment.toJSON())
        }
    }
}

File: colab-backend/mod-calendar-data/src/main/java/org/knollinger/colab/calendar/models/CalendarEventFull.java
package org.knollinger.colab.calendar.models;

import java.util.List;

import org.knollinger.colab.filesys.models.INode;
import org.knollinger.colab.user.models.User;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * 
 */
@Builder()
@NoArgsConstructor()
@AllArgsConstructor()
@Getter(AccessLevel.PUBLIC)
@Setter(AccessLevel.NONE)
public class CalendarEventFull
{
    private CalendarEventCore core;
    private List<User> reqPersons;
    private List<User> optPersons;
    private List<String> hashTags;
    private List<INode> attachments;
}
